Add visitor pattern methods

jakirkham opened the pull request:
Fixes https://github.com/alimanfoo/zarr/issues/92

Provides `visit` and `visititem` that behave identically to their h5py counterparts. This should make it easier to traverse the full hierarchy of a Zarr group. Also should make it easier for users with h5py code to try out using Zarr with fewer changes.

jakirkham commented:
Only has doctests ATM. Could use some unit tests as well. That said, please let me know what you think.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+
+        Examples
+        --------
+        >>> from __future__ import print_function, unicode_literals
```
I don't think this is needed in the example. For Zarr docstrings are standardised to PY3(6).

jakirkham replied:
Yeah, I figured it couldn't hurt.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+        >>> g2 = g1.create_group('foo')
+        >>> g3 = g1.create_group('bar')
+        >>> d1 = g1.create_dataset('baz', shape=100, chunks=10, fill_value=0)
+        >>> d2 = g1.create_dataset('quux', shape=200, chunks=20, fill_value=0)
```
Maybe use 'foo/quux' as the dataset name or something like that, to show what happens when you have a hierarchy?

jakirkham replied:
We can do that. Was thinking about doing something similar. Though the tests explore this as well.

May also drop datasets (or at least disable compression) as the compression levels seem to vary, which would make the doctest annoying to keep.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+        >>> g2 = g1.create_group('foo')
+        >>> g3 = g1.create_group('bar')
+        >>> d1 = g1.create_dataset('baz', shape=100, chunks=10, fill_value=0)
+        >>> d2 = g1.create_dataset('quux', shape=200, chunks=20, fill_value=0)
```
Again suggest to have an example showing what happens with a hierarchy, e.g., 'foo/quux'.

jakirkham replied:
xref: https://github.com/alimanfoo/zarr/pull/122#discussion_r102634014

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
                                  synchronizer=self._synchronizer)
 
+    def visit(self, func):
+        """Run callable on each object name.
```
Suggest to replace ``callable`` with `` `func` ``. Also maybe worth adding a note about how the return value is handled, e.g., from h5py source: "Returning None continues iteration; returning anything else aborts iteration and returns that value."

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+                return value
+
+    def visititems(self, func):
+        """Run callable on each object name and object.
```
As for `visit()`, suggest to replace `callable` with `` `func` `` and adding a note about how the return value is handled.

jakirkham replied:
Note to self: https://github.com/alimanfoo/zarr/pull/122#discussion_r102596944

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+
+        Examples
+        --------
+        >>> from __future__ import print_function, unicode_literals
```
Not needed I don't think.

jakirkham replied:
Same as above.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+        def _visit(obj, func=func):
+            keys = sorted(getattr(obj, "keys", lambda : [])())
+            for each_key in keys:
+                for value in _visit(obj[each_key], func=func):
```
Think there's a call to ``func(each_key)`` missing here?

jakirkham replied:
Yeah, I was playing with stuff and pushed a bad commit. Unfortunately have been AFK until now. Should be fixed.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +416,78 @@ def arrays(self):
+
+        def _visit(obj, func=func):
+            keys = sorted(getattr(obj, "keys", lambda : [])())
+            for each_key in keys:
```
Missing call to ``func(each_key, obj[each_key])``?

jakirkham replied:
Same as above.

jakirkham commented:
Ok, addressed comments and fixed bugs. This is ok on my end.

alimanfoo commented:
Something weird happening with CI, maybe related to install of appdirs? Try upgrading to ``appdirs==1.4.1`` in ``requirements_dev.txt``? Or maybe remove ``appdirs`` altogether from ``requirements_dev.txt`` as it's probably not important to pin it, and let latest get brought in by whichever package requires it?

jakirkham commented:
Yeah, this cropped up with the release of 1.4.1. Have raised upstream to see if they have any idea.

xref: https://github.com/ActiveState/appdirs/issues/89

jakirkham commented:
Put together PR ( https://github.com/alimanfoo/zarr/pull/124 ) to try and fix the CI issues.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +417,89 @@ def arrays(self):
+
+        Examples
+        --------
+        >>> from __future__ import print_function, unicode_literals
```
Same as above.

jakirkham replied:
Also done.

alimanfoo commented on zarr/hierarchy.py:
```diff
@@ -414,6 +417,89 @@ def arrays(self):
+
+        """
+
+        def _visit(obj):
```
This is now a duplicate of _visit above. Suggest to factor out into a single private function.

jakirkham replied:
Yep, was actually thinking of refactoring it differently though. I think we could have a `visitvalues` function that both `visit` and `visititems` use. Have a WIP branch for this, but it isn't quite done yet.

jakirkham replied:
Have pushed a refactoring using `visitvalues`, which seems to work nicely.

jakirkham commented on zarr/hierarchy.py:
```diff
@@ -414,6 +417,89 @@ def arrays(self):
+
+        Examples
+        --------
+        >>> from __future__ import print_function, unicode_literals
```
Sure, done.

jakirkham commented:
Also made changes so that all of these will show up in the docs. Built the docs locally to make sure.

alimanfoo commented:
Thanks for the changes, looks good.

So I'm just reading the [h5py docs](http://docs.h5py.org/en/latest/high/group.html#Group.visit), and apparently the name passed into the visiting function "will be the name of the object relative to the current group". As currently coded, this PR will pass the object path, which is the name relative to the root group. I guess if we are aiming for h5py compatibility where possible then we need to follow the same behaviour? Unless you have a reason to do it differently?

jakirkham commented:
Just to make sure we are on the same page, by current group, I believe they mean the group we called `visit*` on. At least that is what the h5py's `visit*` methods seem to do.

That's a good point. Thanks for raising it. This was definitely an oversight on my part. Not intentional.

Do you have any suggestions for how we should do this? If not, I'm thinking we could just strip out that first `x` characters that correspond to the parent group path or replace the first match.

alimanfoo commented:
No problem, yes that's how I read the docs, name is supposed to be relative to the group we called ``visit*`` on. I think it would be fine to do something like slice the ``.path`` property of the object being visited, using the length of the ``.path`` property on the group we called ``visit*`` on as the slice start coordinate.

jakirkham commented:
Alright, I have now updated the `visit` and `visititems` functions to provide relative paths. Added an example of this to the documentation. Also included a bunch of related unit tests. Hopefully that addresses it.

alimanfoo commented:
Great, thank you!

jakirkham commented:
Thanks for the good review.
